### ISSUE TITLE
Can't use SelenideElement if it was declared before setWebDriver() method

### ISSUE THREAD

**Issue by pavelpp** [selenide/selenide#873]
## The problem
As per issue title, declaring a SelenideElement before calling `WebDriverRunner.setWebDriver(driver)` method leads to an IllegalStateException whenever user tries to interact with the element.

## Details
If we do not use setWebDriver method, then everything works regardless of where you declare SelenideElement. It can be a test class field for instance and still works. Exception stacktrace

```
java.lang.IllegalStateException: Webdriver has been closed. You need to call open(url) to open a browser again.

	at com.codeborne.selenide.drivercommands.LazyDriver.getWebDriver(LazyDriver.java:65)
	at com.codeborne.selenide.impl.ElementFinder.getSearchContext(ElementFinder.java:86)
	at com.codeborne.selenide.impl.ElementFinder.getWebElement(ElementFinder.java:74)
	at com.codeborne.selenide.impl.WebElementSource.checkCondition(WebElementSource.java:47)
	at com.codeborne.selenide.commands.Should.should(Should.java:35)
	at com.codeborne.selenide.commands.Should.execute(Should.java:29)
	at com.codeborne.selenide.commands.Should.execute(Should.java:12)
	at com.codeborne.selenide.commands.Commands.execute(Commands.java:144)
	at com.codeborne.selenide.impl.SelenideElementProxy.dispatchAndRetry(SelenideElementProxy.java:99)
	at com.codeborne.selenide.impl.SelenideElementProxy.invoke(SelenideElementProxy.java:65)
	at com.sun.proxy.$Proxy5.shouldBe(Unknown Source)
```

## Tell us about your environment
* **Selenide Version:** 5.0.1
* **Chrome\Firefox\IE Version:** irrelevant
* **Browser Driver Version:** irrelevant
* **Selenium Version:** 3.x
* **OS Version:** irrelevant

## Code To Reproduce Issue [ Good To Have ]
Easily reproducible with:
```
        SelenideElement a=$("abc");
        WebDriverRunner.setWebDriver(new ChromeDriver());
        open("http://google.com");
        a.shouldNot(exist);
```
And this way it works
```
        WebDriverRunner.setWebDriver(new ChromeDriver());
        SelenideElement a=$("abc");
        open("http://google.com");
        a.shouldNot(exist);
```
Also this works
```
public class SimpleTest {

    public SelenideElement search = $(By.name("q"));

    @Test
    public void testSimple() {
        open("https://www.google.com");
        search.shouldBe(Condition.visible);
    }

}
```
And this does not
```
public class SimpleTest {

    public SelenideElement search = $(By.name("q"));

    @Test
    public void testSimple() {
        WebDriverRunner.setWebDriver(new ChromeDriver(new ChromeOptions()));
        open("https://www.google.com");
        search.shouldBe(Condition.visible);
    }
}
```

**Comment by jkromski** [comment]
From what I understand, by using $() you setting up a default driver and $ uses it to declare SelenideElement, so if you're setting driver manually after $ then Selenide closes previous one and that is why you get the error.
```java
SelenideElement a=$("abc"); // default driver (created for you automatically)
WebDriverRunner.setWebDriver(new ChromeDriver()); // your driver, default driver being closed
open("http://google.com"); // using your driver
a.shouldNot(exist); // as this is proxy it remembers default driver from line 1
```

I think this is not an bug, you should set driver before using it. if you need to declare element before then declare selector only as String and after setting up the driver use $()

**Comment by vinogradoff** [comment]
$() is not supposed to set up any particular driver, it just making use of
a drivers configured at other places.

So I guess it is a bug)

Viele Grüße

Alexei Vinogradov

**Comment by pavelpp** [comment]
@jkromski  that would have been true if it didn't work without using setWebDriver() method, letting Selenide setup driver, but it's not the case. The issue is only happening when setWebDriver method is used. See my example above without usage of setWebDriver, where element is declared as a field - this still works.

**Comment by vinogradoff** [comment]
Actually, I've looked into code and it looks like bug. I hope to fix it
until Sunday.

Viele Grüße

Alexei Vinogradov

**Comment by asolntsev** [comment]
@pavelpp @jkromski  @vinogradoff No, it's not a bug. It was a decision made in Selenide 5.0. See https://selenide.org/2018/10/10/selenide-5.0.0/: 

> Methods $ and $$ now throw an exception if browser is not opened yet (or is already closed). The right way is to open a browser first (by calling open(url)) and then search elements ($, $$ etc).

It may see a bit unexpected, but I believe it will cause a better design of your tests. For me, it's really a strange decision to declare `SelenideElement` as a field in test class. Let's initialise it when it's needed.

**Comment by pavelpp** [comment]
>It may seem a bit unexpected
It is indeed unexpected as it used to work before and not anymore. Secondly, if it's not supposed to work like this then behaviour is not consistent when setWebDriver method is not used. How do you explain this? @asolntsev Make it work the same way in both cases, otherwise it just does not make any sense.

**Comment by asolntsev** [comment]
@pavelpp Well, I agree, Selenide should throw IllegalStateException even if `setWebDriver` was not used. I will do it in version 5.1, thanks for the suggestion. 

P.S. But it will not solve your problem. :)

**Comment by pavelpp** [comment]
@asolntsev it's fine as long as the behaviour is consistent

**Comment by symonk** [comment]
Agree, lets improve what we are telling the user under this circumstance, but I think its a very uncommon to write test code like this

**Comment by vinogradoff** [comment]
I still see it as a bug. I will propose a fix later this week.

**Comment by asolntsev** [comment]
@vinogradoff before you start fixing something, could you explain why you think it's a bug? It was an intentional change.

**Comment by vinogradoff** [comment]
Sure.

This fails.
```
        SelenideElement a=$("abc");
        WebDriverRunner.setWebDriver(new ChromeDriver());
        open("http://google.com");
        a.shouldNot(exist);
```
This  works
```
WebDriverRunner.setWebDriver(new ChromeDriver());
        SelenideElement a=$("abc");
        open("http://google.com");
        a.shouldNot(exist);
```

WebDriver is changed before open - perfectly fine. There is no reason to connect elements definition with any WebDriver instance, so it may happen in any place. And as far as I remember it is not even connected, but doesn't work as a side effect due to some other minor issue in setWebDriver calls.

**Comment by asolntsev** [comment]
And where is the real problem?
Aren't you agree that much better is to first open a browser and then find elements?

**Comment by vinogradoff** [comment]
I still insist that the line doesn't search and doesn't find. It declares location algorithm. Declaration can be placed anywhere. 

This example is of course

**Comment by asolntsev** [comment]
@vinogradoff Who said that? Why? I never said it. `$` just loads elements *lazily*. 

Anyway, my question is still in the air:
* And where is the real problem? (snippet above can be easily fixed: just move `$` below `setWebDriver`. I don't see any problems here.)
* Aren't you agree that much better is to first open a browser and then find elements?

**Comment by vinogradoff** [comment]
There is a better solution - to make a code change and let the both snippets to have the same result.

**Comment by asolntsev** [comment]
You are not answering my questions. :(

**Comment by vinogradoff** [comment]
If you insist.

> Who said that? Why? I never said it

I was never saying it was you. I don't know who says it except me. Why - because it work this way.

> And where is the real problem? 

The different behaviour of the two snippets is illogical. It is a real problem with quality, yes. That you have workaround doesn't make the problem less real.

> Aren't you agree that much better is to first open a browser and then find elements?

We cannot find element before opening the browser, for sure. But even if you put _"find"_ and _"declare"_ to be equal, and even if I would agree that it is better _to declare_ element after `open()` and not before. Just look again at this two snippets. They **both** declare element before `open()`, but one of the is working and the other isn't.


Then please answer my question in return:

Why you resist a change that would make these both pieces of code work consistently? What are the reasons not to do it? (of course you must not do it yourself).

**Comment by vinogradoff** [comment]
> Why are you calling a hack? It's not a hack, it's a totally valid code.

Where I called `By username = By.name("username");` a hack?

> If you want to declare elements, just declare By

Why? Why you don't want me to declare SelenideElement as I always do the last years?

**Comment by vinogradoff** [comment]
> Selenide should throw IllegalStateException in both cases

just beware, that if you throw IllegalStateException every time, $() is used, bevor open() is called - it will break plenty of test, probably my tests too. For what reason?

**Comment by asolntsev** [comment]
@BlackwellQA Well, it was not a bug, it was an intentional change because I believed it will cause better design of tests. :) 
But now I see that it causes a big inconvenience for you (and not only you), so I should probably change my mind. :)

But still, don't you think that having 800+ fields in a single class is a bad design? I mean, BAD design? It breaks all principles of software disciplines like SRP, KISS etc. A big class with many responsibilities is a well-known antipattern in programming (not only in test automation).

Don't you think that splitting the huge class to many small page objects would cause a much better design?

**Comment by Blackwell174** [comment]
@asolntsev Thank you for your reply Andrei :)

I managed to get everything working by adding By and $() everywhere, which is not bad but more things to type when with just elements collection looks a bit cleaner :) The error was very confusing and I tried to rewrite my browser factory and everything, but issue was that elements were before page opening separated from the tests. 

Design wise I understand where you're coming from, but in my specific case I do DNA traits and indexes automation and they all have to be together in one class unfortunately.

**Comment by ivalitov** [comment]
I preapare instances of page objects before running parameterised tests via Juni5 ArgumentsProvider. Page objects already includes fields and actions with $. Before each paremeterised test execution choosing browser.
If browser changes from default, it produses that problem..

**Comment by Skaramus1990** [comment]
Hi, I have the same problem.
I using Spring Boot in my tests. When I run tests in many threads, I have this exception. Only 1 from 5 thread works fine. If I added in my object @Scope(value =ConfigurableBeanFactory.SCOPE_PROTOTYPE, proxyMode = ScopedProxyMode.TARGET_CLASS) all works fine. Or if i write = new MyObject(); before calling element's method. I think problem with the initialization of the elements in singleton object in many treads.

**Comment by Skaramus1990** [comment]
In @Scope the interfaces was added  proxyMode = ScopedProxyMode.TARGET_CLASS. It solved my problem.

EXAMPLE:

Target (TYPE)
@Retention (RetentionPolicy.RUNTIME)
@Component
@Scope ("thread", **proxyMode = ScopedProxyMode.TARGET_CLASS**)
@Inherited
public @interface PageObject {}

**Comment by MichalUrbaniakQA** [comment]
I have the same problem:

@Before(order = 1)
    public void initProperties() throws ReadPropertiesFileException {
        propertiesConfig = new PropertiesConfig();
    }

    @Before(order = 2)
    public void setConfiguration() {
        Configuration.fastSetValue = true;
        Configuration.screenshots = true;
        Configuration.timeout = 15000;
        Configuration.reportsFolder = "reports/screens";
    }

    @Before(order = 3)
    public void openUrl() {
        System.setProperty("webdriver.gecko.driver", "src/test/resources/drivers/geckodriver0240.exe");
        WebDriverRunner.setWebDriver(new FirefoxDriver());
        open(propertiesConfig.getDemoUrl());
        getWebDriver().manage().deleteAllCookies();
}

After(in different class) I want to use:

public class LoginPage {
    private String navbarSelector = ".nav.navbar-nav.hidden-sm.go-left";
    private SelenideElement myAccountDropdown = $(navbarSelector).$$("li").first();
    private SelenideElement loginOption = $(navbarSelector).$(".dropdown-menu").$$("li").first().$(".go-text-right");
    private SelenideElement demoUserEmail = $x("//input[@type='email']");
    private SelenideElement demoUserPassword = $x("//input[@type='password']");

    public LoginPage openMyAccountDropdown() {
        myAccountDropdown.click();

        return this;
    }


And I am getting : 

`java.lang.IllegalStateException: Webdriver has been closed. You need to call open(url) to open a browser again.
	at com.codeborne.selenide.drivercommands.LazyDriver.getWebDriver(LazyDriver.java:64)
	at com.codeborne.selenide.impl.ElementFinder.getSearchContext(ElementFinder.java:86)
	at com.codeborne.selenide.impl.ElementFinder.getWebElement(ElementFinder.java:74)
	at com.codeborne.selenide.impl.SelenideElementProxy.dispatchAndRetry(SelenideElementProxy.java:102)
	at com.codeborne.selenide.impl.SelenideElementProxy.invoke(SelenideElementProxy.java:65)
	at com.sun.proxy.$Proxy4.findElements(Unknown Source)
	at com.codeborne.selenide.impl.WebElementSelector.findElements(WebElementSelector.java:37)
	at com.codeborne.selenide.impl.BySelectorCollection.getElements(BySelectorCollection.java:30)
	at com.codeborne.selenide.impl.CollectionElement.getWebElement(CollectionElement.java:35)
	at com.codeborne.selenide.impl.WebElementSource.checkCondition(WebElementSource.java:50)
	at com.codeborne.selenide.impl.WebElementSource.findAndAssertElementIsInteractable(WebElementSource.java:87)
	at com.codeborne.selenide.commands.Click.execute(Click.java:13)
	at com.codeborne.selenide.commands.Click.execute(Click.java:9)
	at com.codeborne.selenide.commands.Commands.execute(Commands.java:144)
	at com.codeborne.selenide.impl.SelenideElementProxy.dispatchAndRetry(SelenideElementProxy.java:99)
	at com.codeborne.selenide.impl.SelenideElementProxy.invoke(SelenideElementProxy.java:65)
	at com.sun.proxy.$Proxy4.click(Unknown Source)
	at pageobject.login.LoginPage.openMyAccountDropdown(LoginPage.java:23)
	at glue.GenericDef.lambda$new$0(GenericDef.java:12)
	at ✽.I am logged to app(file:/D:/Workspace/intelij/phptravels/src/test/resources/features/smoke/smokeLanguage.feature:8)`

**Comment by asolntsev** [comment]
@pavelpp @jkromski @BlackwellQA @ivalitov  @Skaramus1990 @MichalUrbaniakQA Congratulations, this change was implemented in [Selenide 5.4.1](https://selenide.org/2019/10/16/selenide-5.4.0/).
